### PR TITLE
Added Moshi component serializer

### PR DESCRIPTION
**Pull request by BomBardyGamer**
As the title states, this PR adds a new component serializer using the [Moshi](https://github.com/squareup/moshi) JSON library. 
This serializer borrows a bit from Gson's code, as Moshi and Gson are similar in parts with the way they function.

I have also copied all of the tests over from the Gson serializer, as I think it is fair that this serializer should be able to at least do everything that the Gson one does, and they are all passing (I ran them all multiple times).

You may also notice that, as per @zml2008 's request, I have included some JMH benchmarks that I copied from [this PR](https://github.com/KyoriPowered/adventure/pull/373). I am running the tests as I am writing this, so I am unable to produce any results yet. Feel free to run these benchmarks yourself and post your results here, as it will help in determining performance differences.

I am also looking to implement more serializers for other JSON libraries in the near future, as I believe it would be useful for people to be able to natively serialize using a library of their choice, rather than being forced to use Gson. These will likely be kept separate from the main project, however, as I think that too many choices can confuse people a bit (unless you think that it is worth getting them merged).

### CODE REVIEW

File: text-serializer-moshi/src/jmh/java/net/kyori/adventure/text/serializer/moshi/ComponentDeserializationBenchmark.java
Comment (kezz) [inline text-serializer-moshi/src/jmh/java/net/kyori/adventure/text/serializer/moshi/ComponentDeserializationBenchmark.java:1]: Missing license

File: text-serializer-moshi/src/jmh/java/net/kyori/adventure/text/serializer/moshi/ComponentSerializationBenchmark.java
Comment (kezz) [inline text-serializer-moshi/src/jmh/java/net/kyori/adventure/text/serializer/moshi/ComponentSerializationBenchmark.java:1]: Missing license

File: text-serializer-moshi/src/jmh/java/net/kyori/adventure/text/serializer/moshi/ComponentDeserializationBenchmark.java
Comment (kezz) [inline text-serializer-moshi/src/jmh/java/net/kyori/adventure/text/serializer/moshi/ComponentDeserializationBenchmark.java:21]: No static imports - this applies elsewhere.

**Review by kezz — CHANGES_REQUESTED** [review]
Just had a brief glance over, a few minor comments. Don't forget to run a style check :p

Some initial questions I have:
1. Do we want to consider a generic `JsonComponentSerializer` to centralise the provider logic as well as allowing implementations/etc to hide the specific JSON library being used?
2. Does this PR produce the **exact** same output as the GSON one?

**Comment by BomBardyGamer** [comment]
> Just had a brief glance over, a few minor comments. Don't forget to run a style check :p
> 
> Some initial questions I have:
> 
>     1. Do we want to consider a generic `JsonComponentSerializer` to centralise the provider logic as well as allowing implementations/etc to hide the specific JSON library being used?
> 
>     2. Does this PR produce the **exact** same output as the GSON one?

I did run a style check... before I added the JMH benchmarks lol

1. That actually could be a really good idea.
2. I'm at least 99.9% sure that it does.

File: text-serializer-moshi/src/main/java/net/kyori/adventure/text/serializer/moshi/StyleAdapter.java
Comment (KingOfSquares) [inline text-serializer-moshi/src/main/java/net/kyori/adventure/text/serializer/moshi/StyleAdapter.java:175]: Why not `decoration.toString()`?

File: text-serializer-moshi/src/main/java/net/kyori/adventure/text/serializer/moshi/StyleAdapter.java
Comment (KingOfSquares) [inline text-serializer-moshi/src/main/java/net/kyori/adventure/text/serializer/moshi/StyleAdapter.java:111]: Same here, `toString()`?

**Review by KingOfSquares — COMMENTED** [review]
If you are moving modules you may want to add a note in the `package-info` files as well(deprecation)?

File: text-serializer-gson/src/main/java/net/kyori/adventure/text/serializer/gson/GsonComponentSerializerImpl.java
Comment (BomBardyGamer) [reply text-serializer-gson/src/main/java/net/kyori/adventure/text/serializer/gson/GsonComponentSerializerImpl.java:158]: Look at the above one, this is the implementation for the deprecated method.

File: text-serializer-moshi/src/main/java/net/kyori/adventure/text/serializer/moshi/StyleAdapter.java
Comment (BomBardyGamer) [reply text-serializer-moshi/src/main/java/net/kyori/adventure/text/serializer/moshi/StyleAdapter.java:175]: This is what is used by the Gson serializer, and so I was following that.

File: text-serializer-moshi/src/main/java/net/kyori/adventure/text/serializer/moshi/StyleAdapter.java
Comment (BomBardyGamer) [reply text-serializer-moshi/src/main/java/net/kyori/adventure/text/serializer/moshi/StyleAdapter.java:111]: Same as above

File: text-serializer-gson/src/main/java/net/kyori/adventure/text/serializer/gson/GsonComponentSerializerImpl.java
Comment (KingOfSquares) [reply text-serializer-gson/src/main/java/net/kyori/adventure/text/serializer/gson/GsonComponentSerializerImpl.java:158]: Oooh, stupid me thought the full qualification was for the annotation 🤦

File: text-serializer-moshi/src/main/java/net/kyori/adventure/text/serializer/moshi/MoshiComponentSerializerImpl.java
Comment (zml2008) [inline text-serializer-moshi/src/main/java/net/kyori/adventure/text/serializer/moshi/MoshiComponentSerializerImpl.java:27]: Don't use internal API here.

**Review by zml2008 — CHANGES_REQUESTED** [review]
It looks like a lot of the adapters here are still using tree-style API (calling `readJsonValue()` directly). Take a look at Emily's PR #373 for a conversion of the GSON serializer to GSON's streaming API -- that's what you should prefer where possible for better performance and reduced allocations.

File: text-serializer-moshi/src/main/java/net/kyori/adventure/text/serializer/moshi/MoshiComponentSerializerImpl.java
Comment (BomBardyGamer) [reply text-serializer-moshi/src/main/java/net/kyori/adventure/text/serializer/moshi/MoshiComponentSerializerImpl.java:27]: Not much we can really do here except write `Util.typesMatch` ourselves, which is probably not the best idea. The reason why we use this is to mimic the behaviour of regular type adapter factories, so we can gain access to the moshi instance to construct the adapter, as otherwise, we cannot construct the adapter because we are still building the Moshi serializer when we need it.

**Comment by BomBardyGamer** [comment]
> It looks like a lot of the adapters here are still using tree-style API (calling `readJsonValue()` directly). Take a look at Emily's PR #373 for a conversion of the GSON serializer to GSON's streaming API -- that's what you should prefer where possible for better performance and reduced allocations.

The reason why I chose to use the tree-style API over the streaming API is to avoid the necessity of the elements being in a specific defined order. I can change this, however, if this is not desired behaviour.

**Comment by kashike** [comment]
Why moshi?

File: text-serializer-gson-legacy-impl/src/main/java/net/kyori/adventure/text/serializer/gson/legacyimpl/NBTLegacyHoverEventSerializer.java
Comment (kezz) [inline text-serializer-gson-legacy-impl/src/main/java/net/kyori/adventure/text/serializer/gson/legacyimpl/NBTLegacyHoverEventSerializer.java:36]: If you're deprecating stuff for removal you can use `@ApiStatus.ScheduledForRemoval`. This applies throughout.

File: text-serializer-gson-legacy-impl/src/main/java/net/kyori/adventure/text/serializer/gson/legacyimpl/NBTLegacyHoverEventSerializer.java
Comment (kezz) [inline text-serializer-gson-legacy-impl/src/main/java/net/kyori/adventure/text/serializer/gson/legacyimpl/NBTLegacyHoverEventSerializer.java:34]: For consistency with the rest of the API. This applies throughout.
```suggestion
 * @deprecated for removal since 4.9.0, use {@link net.kyori.adventure.text.serializer.json.legacyimpl.NBTLegacyHoverEventSerializer} instead.
```

File: text-serializer-gson-legacy-impl/src/main/java/net/kyori/adventure/text/serializer/gson/legacyimpl/NBTLegacyHoverEventSerializer.java
Comment (kezz) [inline text-serializer-gson-legacy-impl/src/main/java/net/kyori/adventure/text/serializer/gson/legacyimpl/NBTLegacyHoverEventSerializer.java:44]: Provide `@deprecated` docs in the javadocs for all elements that are deprecated. This applies throughout.

File: text-serializer-json/src/main/java/net/kyori/adventure/text/serializer/json/Instances.java
Comment (kezz) [inline text-serializer-json/src/main/java/net/kyori/adventure/text/serializer/json/Instances.java:30]: Are you even using `orElseGet`? Not sure what this is referring to.

**Review by kezz — CHANGES_REQUESTED** [review]
I like this idea of a Moshi serializer, it's definitely faster so for those who are happy with the bulky Kotlin dependency (or who use it anyway) it's probably a good alternative to Gson. However, I'm a bit unhappy with the scope of this PR in general. I like the idea of a generic `JsonComponentSerializer` but I think that including it in this PR alongside a Moshi serializer is a bit noisy and pollutes the scope of this PR. I'd _much_ prefer them to be split into two separate PRs so that time can be spent reviewing them both separately. 

Additionally I've left several minor review comments for consideration.

File: text-serializer-json/src/main/java/net/kyori/adventure/text/serializer/json/Instances.java
Comment (BomBardyGamer) [reply text-serializer-json/src/main/java/net/kyori/adventure/text/serializer/json/Instances.java:30]: Look in `JsonComponentSerializer`. The static accessors use `orElseGet`. That's what I mean.

Comment (kezz) [reply text-serializer-json/src/main/java/net/kyori/adventure/text/serializer/json/Instances.java:30]: They use `orElseThrow`?

Comment (BomBardyGamer) [reply text-serializer-json/src/main/java/net/kyori/adventure/text/serializer/json/Instances.java:30]: My bad, I thought they used `orElseGet` lol, I can change this commit if necessary.

**Comment by BomBardyGamer** [comment]
Common JSON abstraction has been moved to #424 , I will remove it from this PR, but I am declaring that this PR is blocked until that other PR is merged, due to me not wanting to add things that will be immediately deprecated.

**Comment by NicoNekoDev** [comment]
I will just leave this here:
https://discord.com/channels/452518336627081236/718515639253205043/936561607650185256
If anyone would appreciate the bounty.

**Comment by zml2008** [comment]
Thank you for your time on this pull request. Unfortunately, in the interests of manging the scope of the adventure project, we aren't able to justify adding an implementation for Moshi at this time.

We will be happy to consider the changes in #424 during the 4.11.0 cycle though. Hopefully this will allow you to maintain the Moshi serializer as a standalone library, as well as support anyone else who wants to implement a component serializer for any other JSON library. If you choose to go ahead with this as an external library, we would also encourage you to contribute it to the [Community Libraries](https://docs.adventure.kyori.net/community-libraries.html) page in our documentation once you feel it's ready, so that the community can learn more about it.

If you run across any other additions or changes that need to be made to adventure's API or the JSON serializer abstraction while working on the Moshi serializer, we'll be happy to consider those contributions as well -- just open an issue or ask in our Discord first.

**Pull request closed without merging**
